### PR TITLE
Update RPO's padding rule to use that in the xHash paper

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,8 +1,3 @@
-## 0.10.0 (TBD)
-
-* Added more `RpoDigest` and `RpxDigest` conversions (#311).
-* Migrated to Winterfell v0.9.0 (#315)
-
```
Comment (bobbinth) [inline CHANGELOG.md:5]: We should probably rebase this PR from the latest `next` to make sure we carry over the latest updates.

File: src/hash/rescue/rpo/mod.rs
Comment (bobbinth) [inline src/hash/rescue/rpo/mod.rs:1]: Should we also update the `hash()` method (starting on line 65)? I forget now why we use different domain separation for hashing bytes vs. hashing elements. Is it not possible to keep the same domains for both (based only on the number of elements) and always pad the last 7-byte chunk with with a byte containing $1$?

Comment (Al-Kindi-0) [reply src/hash/rescue/rpo/mod.rs:1]: Ah, yes! We should probably use the one from RPX. We used an extra domain separation to separate hashing bytes from hashing field elements.

Comment (bobbinth) [reply src/hash/rescue/rpo/mod.rs:1]: > We used an extra domain separation to separate hashing bytes from hashing field elements.

Right - but I forget why we need to do that. Wouldn't applying the same rule consistently be simpler?

Comment (Al-Kindi-0) [reply src/hash/rescue/rpo/mod.rs:1]: The reason was so that we avoid collisions, if I remember correctly.

Comment (bobbinth) [reply src/hash/rescue/rpo/mod.rs:1]: Do you remember what kind of collisions could happen here (or why that would be a problem)? Maybe we should leave a note about this in the comments as I think we did discuss it be I can't remember what the problem was.

Comment (Al-Kindi-0) [reply src/hash/rescue/rpo/mod.rs:1]: Mainly that a sequence of bytes (i.e., the first domain) could hash to the same output as a sequence of field elements (i.e., the second domain). Maybe it is not an issue for the use cases of RPO∕RPX but I believe it is standard practice to make sure that sequences of bytes coming from different sources/domains are hashed to different outputs.

Comment (bobbinth) [reply src/hash/rescue/rpo/mod.rs:1]: Let's copy over the padding methodology we use for RPX here.

Comment (Al-Kindi-0) [reply src/hash/rescue/rpo/mod.rs:1]: Done!
When doing that, I discovered a bug in that implementation for RPX. I fixed it now and added tests to RPX. The only test missing is the test-vectors, which we can comeback to later.
There was also the issue of hashing an empty `Vec<u8>`. I have added a fix, though not the cleanest, should be sufficient to solve it.

File: src/hash/rescue/rpx/mod.rs [hunk 2]
```diff
@@ -88,9 +91,16 @@ impl Hasher for Rpx256 {
...
         // `i` is not zero, then the chunks count wasn't enough to fill the state range, and an
         // additional permutation must be performed.
+        let mut current_element = 0_usize;
+        // handle the case of an empty `bytes`
+        let last_element = if num_field_elem == 0 {
+            current_element
+        } else {
+            num_field_elem - 1
+        };
         let i = bytes.chunks(BINARY_CHUNK_SIZE).fold(0, |i, chunk| {
```
Comment (bobbinth) [inline src/hash/rescue/rpx/mod.rs:101]: I would also probably rename `i` into something like `rate_pos` to make the naming explicit.

File: src/hash/rescue/rpx/mod.rs [hunk 3]
```diff
@@ -88,9 +91,16 @@ impl Hasher for Rpx256 {
...
             // copy the chunk into the buffer
-            if i != num_field_elem - 1 {
+            if current_element != last_element {
```
Comment (bobbinth) [inline src/hash/rescue/rpx/mod.rs:103]: Could we change this condition into something like:
```Rust
if chunk.len() == BINARY_CHUNK_SIZE {
```
Then, we won't need current/last element variables (and the tests seems to pass).

**Review by bobbinth — COMMENTED** [review]
Great catch! I've added a few more comments to clarify the naming and potentially simplify things a bit.

File: src/hash/rescue/rpx/mod.rs (hunk 2 above)
Comment (Al-Kindi-0) [reply src/hash/rescue/rpx/mod.rs]: Changed

File: src/hash/rescue/rpx/mod.rs (hunk 3 above)
Comment (Al-Kindi-0) [reply src/hash/rescue/rpx/mod.rs]: I went ahead and changed but then realized that we actually want to pad with a 1 always. The case of having a final chunk which is of size 7 would not be covered by the above change. 
I added a test to illustrate when things break.

Comment (bobbinth) [reply src/hash/rescue/rpx/mod.rs]: Great catch!

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! Let's hold off on merging this until we propagate the changes to `miden-base`.

On potential thing left to consider: I believe currently hashing of empty list of elements and hashing of 8 `ZERO` elements will give us the same result - right? I wonder if there is a good way to fix this - just in case. Nothing immediately comes to mind though.

**Comment by Al-Kindi-0** [comment]
> On potential thing left to consider: I believe currently hashing of empty list of elements and hashing of 8 `ZERO` elements will give us the same result - right? I wonder if there is a good way to fix this - just in case. Nothing immediately comes to mind though.

I think we had a similar discussion in RPO and I believe we settled on not accepting empty lists for hashing. I will think about it again to see if we can accommodate it differently.

**Comment by bobbinth** [comment]
@Al-Kindi-0 - let's rebase this PR.

Also, should we implement the rule where hashing of an empty string gives us `[ZERO; 4]` output?

**Comment by Al-Kindi-0** [comment]
> @Al-Kindi-0 - let's rebase this PR.
> 
> Also, should we implement the rule where hashing of an empty string gives us `[ZERO; 4]` output?

Done!

File: src/hash/rescue/rpx/mod.rs [hunk 4]
```diff
@@ -68,6 +71,9 @@ pub type CubicExtElement = CubeExtension<Felt>;
 /// the bottleneck for the security bound of the sponge in overwrite-mode only when it is
 /// lower than 2^128, we see that the target 128-bit security level is maintained as long as
 /// the size of the domain identifier space, including for padding, is less than 2^128.
+///
+/// ## Hashing of empty lists
+/// The current implementation hashes empty lists to the zero digest [0, 0, 0, 0].
```
Comment (bobbinth) [inline src/hash/rescue/rpx/mod.rs:76]: I would probably call this "Hashing empty input" or maybe "Hashing empty strings" and would also expand this a bit to mention the motivation - i.e., on empty input we do not apply a permutation.

File: src/hash/rescue/rpo/mod.rs [hunk 5]
```diff
@@ -64,6 +65,9 @@ mod tests;
 /// becomes the bottleneck for the security bound of the sponge in overwrite-mode only when it is
 /// lower than 2^128, we see that the target 128-bit security level is maintained as long as
 /// the size of the domain identifier space, including for padding, is less than 2^128.
+///
+/// ## Hashing of empty lists
+/// The current implementation hashes empty lists to the zero digest [0, 0, 0, 0].
 #[derive(Debug, Copy, Clone, Eq, PartialEq)]
 pub struct Rpo256();
```
Comment (bobbinth) [inline src/hash/rescue/rpo/mod.rs:72]: Same comment as above.

File: src/hash/rescue/rpx/tests.rs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    // ----------------------------------------------------------
+
+    // set domain to ZERO. This should not change the result.
+    let domain = ZERO;
+
+    let merge_in_domain_result = Rpx256::merge_in_domain(&digests, domain);
+    assert_eq!(merge_result, merge_in_domain_result);
+
+    // ------------- merge with domain = 1
+    // ----------------------------------------------------------
```
Comment (bobbinth) [inline src/hash/rescue/rpx/tests.rs:43]: Let's shorten this separator to avoid wrapping.

File: src/hash/rescue/rpx/tests.rs [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    let elements = [Felt::new(rand_value()); 8];
+
+    let digests: [RpxDigest; 2] = [
+        RpxDigest::new(elements[..4].try_into().unwrap()),
+        RpxDigest::new(elements[4..].try_into().unwrap()),
+    ];
+    let merge_result = Rpx256::merge(&digests);
+
+    // ------------- merge with domain = 0
+    // ----------------------------------------------------------
```
Comment (bobbinth) [inline src/hash/rescue/rpx/tests.rs:34]: Same comment as above.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left a few small comments inline.

Also, it seems like the "return default digest on empty input" as already working here, right?

Lastly, could you refresh https://github.com/0xPolygonMiden/miden-vm/pull/1343 when you get a chance?

File: src/hash/rescue/rpx/mod.rs (hunk 4 above)
Comment (Al-Kindi-0) [reply src/hash/rescue/rpx/mod.rs:76]: Changed

File: src/hash/rescue/rpo/mod.rs (hunk 5 above)
Comment (Al-Kindi-0) [reply src/hash/rescue/rpo/mod.rs:72]: Changed

File: src/hash/rescue/rpx/tests.rs (hunk 6 above)
Comment (Al-Kindi-0) [reply src/hash/rescue/rpx/tests.rs:43]: Fixed

File: src/hash/rescue/rpx/tests.rs (hunk 7 above)
Comment (Al-Kindi-0) [reply src/hash/rescue/rpx/tests.rs:34]: Fixed

**Comment by Al-Kindi-0** [comment]
> Also, it seems like the "return default digest on empty input" as already working here, right?
> 

Yes, had to check the bytes case, but even then it worked without complications

> Lastly, could you refresh [0xPolygonMiden/miden-vm#1343](https://github.com/0xPolygonMiden/miden-vm/pull/1343) when you get a chance?

Sure thing

**Pull request merged**